### PR TITLE
Reimplemented drag-n-drop functionality

### PR DESCRIPTION
**Pull request by TheQwertiest**
**TLDR:**
- Implemented IDropSource interface: now it's possible to drag items from JScript to other panels or even outside fb2k.
- Reimplemented IDropTarget interface:
  - Does not override allowed effects from DoDragDrop.
  - Handles effects from panel, which also enables use of native cursor
changing.
  - Removed unneeded fields and methods: Action.ToPlaylist and Action.Mode.
- Changed on_drag_enter/on_drag_over/on_drag_drop interface handling for consistency with it's WinAPI counter part: allowed effects are now supplied via Action.Effect and the handler MUST return the effect it needs through the same field.
- Added fb.DoDragDrop method to initiate drag from panel.

**Documentation:**
```
/// @details This methods blocks execution of callee function (all callbacks will be called properly though).
///                  Drag will be canceled if either escape button or other mouse buttons are pressed.
/// @param[in] items Handle list to drag  
/// @param[in] okEffects Allowed effects for drop operation (for possible values see https://msdn.microsoft.com/en-us/library/windows/desktop/ms693457.aspx)  
/// @param[out] p Pointer to effect from on_drag_drop or DROPEFFECT_NONE if drag-n-drop was cancelled.  
fb.DoDragDrop( IFbMetadbHandleList* items, UINT okEffects, [out, retval] UINT* p );
```

___
**Examples:**  

**Initiate drag from panel:**
```
// Initiates drag-n-drop of the selected items. Allowed actions are COPY, MOVE and LINK
var effect = fb.DoDragDrop(plman.GetPlaylistSelectedItems(plman.ActivePlaylist), 1|2|4 );
// Someone handled drop with effect DROPEFFECT_COPY
console.log(effect) // 1
```

**Handle drag start:**
```
function on_drag_enter (action, x, y, mask) {
    if (plman.IsPlaylistLocked(cur_playlist_idx)) {
        action.Effect = 0; // DROPEFFECT_NONE
    }
    else {
        action.Effect = (action.Effect & 1) || (action.Effect & 2) || (action.Effect & 4);
    }
    // do stuff...
};
```

### CODE REVIEW

**Comment by TheQwertiest** [comment]
Potentially in can be expanded to allow other types of data to be dragged, like urls, file paths, images and etc,

**Comment by marc2k3** [comment]
Cool. I'm in no state to test this now but I'll get on it tomorrow.

File: src/script_interface.h [hunk 1]
```diff
@@ -446,13 +447,14 @@ __interface IThemeManager : IDisposable
 ]
 __interface IDropSourceAction : IDisposable
 {
-	STDMETHOD(ToPlaylist)();
 	[propget] STDMETHOD(Parsable)([out, retval] VARIANT_BOOL* parsable);
```
Comment (marc2k3) [inline src/script_interface.h]: Does Parsable really need to be a property accessible via script?

File: src/script_interface_impl.cpp [hunk 2]
```diff
@@ -2012,6 +2035,30 @@ STDMETHODIMP FbUtils::CreateProfiler(BSTR name, IFbProfiler** pp)
 	return S_OK;
 }
 
+STDMETHODIMP FbUtils::DoDragDrop(IFbMetadbHandleList* items, UINT okEffects, UINT* p)
+{
+	if ( !p || !items ) return E_POINTER;
```
Comment (marc2k3) [inline src/script_interface_impl.cpp]: I don't think you need to check `items` there as the script already errors if you supply invalid params. Also, I'd like to see the parentheses/spacing of arguments match the existing style. I don't mind tidying this up if you can't be bothered but please consider it in future!

**Comment by TheQwertiest** [comment]
- Cleaned up the code.
- Removed Parseable field altogether, since allowed actions are regulated through Effects.
- fb2k allowed effect are now saved on enter and applied on each drag operation, since pdwEffect argument value is always equal to dwOkEffects from DoDragDrop.

**Comment by TheQwertiest** [comment]
PS: As always - I will squash PR into a single commit, once it is approved.

**Comment by marc2k3** [comment]
No need. Github can squash while merging.

Thanks again for your efforts.

**Comment by TheQwertiest** [comment]
Oh, that's neat, didn't know that =)
Thanks!

**Comment by marc2k3** [comment]
I need some help!! I think I can figure out how to do some examples for dragging in/out but what's the best way to handle moving tracks within the same panel?

**Comment by TheQwertiest** [comment]
Well, I could not find a way to make an in-panel drag-n-drop utilizing only native drag-n-drop mechanic :\
Nor did I manage to think of a JScript API extension, that could've helped with that.
The problem is that we need to handle drop differently depending on the drop location: drops from outside to panel and from panel to outside are handled natively by copying items; but if it is in-panel with Move effect, then we need only to move items (and handling it by copying then removing the leftovers is kinda wasteful).

In my theme I'm using native drag-n-drop API for dragging in and out, but intra panel drag-n-drop is mostly handled manually: 
1. I set a flag, when I'm the one who's initiating the drag (i.e. fb.DoDragDrop): https://github.com/TheQwertiest/CaTRoX_QWR/blob/master/theme/Scripts/Panel_Playlist.js#L3626
2. All the drag-n-drop is handled the same way as external drag-n-drop.
3. On on_drag_drop I'm checking for the flag value: if it is not set, I use native drop, but if it is - I cancel native drop and handle it manually by moving tracks around:  
https://github.com/TheQwertiest/CaTRoX_QWR/blob/master/theme/Scripts/Panel_Playlist.js#L1003
and
https://github.com/TheQwertiest/CaTRoX_QWR/blob/master/theme/Scripts/Panel_Playlist.js#L3737
